### DISCUSSION TITLE
Playback of ESP embedded audio in conjunction with BluetoothA2DPSink

### DISCUSSION THREAD

**Discussion by KiShiVi** [pschatzmann/ESP32-A2DP#579] [Ideas]
Good afternoon, I was very interested in playing back audio stored in FLASH memory using BluetoothA2DPSink. I managed to solve my problem as follows:

1) In BluetoothA2DPSink.h I changed the access modifier of ```bool is_output = true;``` from protected to public (which is very bad).
2) In BluetoothA2DPSink.h I changed the access modifier of ```virtual size_t write_audio(const uint8_t *data, size_t size)``` from protected to public.
3) In BluetoothA2DPSink.cpp in the function ```size_t BluetoothA2DPSink::i2s_write_data(const uint8_t *data, size_t item_size)``` I commented out the check ```if (!is_output) return item_size;```. 

This way I managed to get my own audio playback (even without AudioTools). 

```cpp
a2dp_sink.is_output = false;
a2dp_sink.write_audio(StarWars30_raw, StarWars30_raw_len);
a2dp_sink.is_output = true;
```

(StarWars30.h from your project's examples was taken as an example).

This code works simultaneously with ```a2dp_sink.start("name")``` and does its job perfectly. The Bluetooth sound is muted, then a fragment of audio is played back, then the Bluetooth sound is resumed.

Is it possible to modify your library in such a simple way? So that it would be possible to play your audio by interfering with the Bluetooth stream (without conflicts).

I would create Pull Request myself, but I'm sure that I may not know many nuances and consequences of my modifications (especially since they are not quite correct and made for the sake of experimentation). Thank you very much!

**Comment by KiShiVi** [comment]
Also, during testing, I found that when I turn on the ESP32 for the first time BEFORE establishing a Bluetooth connection, my method works.  After connecting via Bluetooth and while playing something via Bluetooth, my method also works.  But as soon as I pause the music (without breaking the connection), my method stops working and the recorded signal does not play back. 
 ```W (213932) BT_AV: i2s_write_data failed - inactive```
 I understand perfectly well that there is another check inside that detects the presence of a signal via Bluetooth and, if it is absent, stops playing EVERYTHING, but I am writing all this to the fact that probably when the controller is turned on for the first time, the I2S output should also be blocked (but it is open)  .  Perhaps you would be interested

**Comment by pschatzmann** [comment]
I don't quite understand: what is the point of this ? 

Why don't you use [a i2s.write()](https://github.com/pschatzmann/ESP32-A2DP?tab=readme-ov-file#a-simple-i2s-example-a2ds-sink-using-default-pins) directly ?

**Reply by KiShiVi** [reply]
As far as I understand, this is not able to stop the receiver ring buffer.  Wouldn't there be a situation where two sources would try to access i2s at once?  Will there be a conflict?

**Comment by pschatzmann** [comment]
Currently you can call e.g. set_stream_reader(nullptr, flag); to update the i2s_output flag.
I have added the new method set_output_active(flag) to make it a bit more explicit...

**Reply by KiShiVi** [reply]
Thank you very much!  I am very grateful for your prompt feedback!
